<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.9.17">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2026-01-22">
<meta name="description" content="I sequenced the genome of the pyoverdine-producing Pseudomonas from a previous post, and had some fun digging into the genes to figure out what molecule exactly it makes">

<title>Sequencing Pseudomonas putida + Bioinformatics &amp; a Novel Pyoverdine Variant (DRAFT) â€“ johnowhitaker.dev</title>
<style>
/* Default styles provided by pandoc.
** See https://pandoc.org/MANUAL.html#variables-for-html for config info.
*/
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark-dcd9e8c573452b099f942d625f6f4e16.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-7808d31791051cc8da4c7717a821f40e.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
<meta name="twitter:title" content="Sequencing Pseudomonas putida + Bioinformatics &amp; a Novel Pyoverdine Variant (DRAFT) â€“ johnowhitaker.dev">
<meta name="twitter:description" content="I sequenced the genome of the pyoverdine-producing Pseudomonas from a previous post, and had some fun digging into the genes to figure out what molecule exactly it makes">
<meta name="twitter:image" content="https://johnowhitaker.dev/posts/images/pseud_contig.png">
<meta name="twitter:creator" content="@johnowhitaker">
<meta name="twitter:card" content="summary">
<meta name="twitter:image-height" content="835">
<meta name="twitter:image-width" content="836">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../all.html"> 
<span class="menu-text">Everything Feed</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-more" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">More</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-more">    
        <li>
    <a class="dropdown-item" href="../tsbabn.html">
 <span class="dropdown-text">TSBABN</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../resume.html">
 <span class="dropdown-text">Resume: Jonathan Whitaker</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../books.html">
 <span class="dropdown-text">Book Recommendations</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../blogs.html">
 <span class="dropdown-text">Blog Recommendations</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../yt.html">
 <span class="dropdown-text">YouTube Channel Recommendations</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../appearances.html">
 <span class="dropdown-text">Appearances</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../things.html">
 <span class="dropdown-text">Thing Recommendations</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://johnowhitaker.dev/index.xml"><i class="bi bi-rss" role="img">
</i> 
 <span class="dropdown-text">RSS Feed</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../now.html">
 <span class="dropdown-text">Now</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#sequencing-with-plasmidsaurus" id="toc-sequencing-with-plasmidsaurus" class="nav-link active" data-scroll-target="#sequencing-with-plasmidsaurus">Sequencing with Plasmidsaurus</a></li>
  <li><a href="#pyoverdine-investigation" id="toc-pyoverdine-investigation" class="nav-link" data-scroll-target="#pyoverdine-investigation">Pyoverdine Investigation</a></li>
  <li><a href="#future-plans" id="toc-future-plans" class="nav-link" data-scroll-target="#future-plans">Future Plans</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Sequencing Pseudomonas putida + Bioinformatics &amp; a Novel Pyoverdine Variant (DRAFT)</h1>
  <div class="quarto-categories">
    <div class="quarto-category">bio</div>
  </div>
  </div>

<div>
  <div class="description">
    I sequenced the genome of the pyoverdine-producing Pseudomonas from a previous post, and had some fun digging into the genes to figure out what molecule exactly it makes
  </div>
</div>


<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 22, 2026</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p>In my <a href="https://johnowhitaker.dev/posts/pseudomonas.html">previous post</a> I showed a bacteria isolated from the water around the roots of a Jade plant on our windowsill, which secreted something that fluoresced a lovely blue under UV light. My best guess was that it was some species of Pseudomonas, and the blue was from pyoverdine. But how can we know for sure? Well, I put it off because money, but since I like working with this bacteria and want to do more with it, I figured Iâ€™d bite the bullet and pay the ~$120 to get my answers :) It turns out that the closest match is P. putida, and that the pyoverdine variant that this makes is probably different to anything documented in the literature. Exciting stuff!</p>
<section id="sequencing-with-plasmidsaurus" class="level2">
<h2 class="anchored" data-anchor-id="sequencing-with-plasmidsaurus">Sequencing with Plasmidsaurus</h2>
<p>I went with Plasmidsaurusâ€™ â€˜Standard Bacterial Genome Sequencing with Extractionâ€™ <a href="https://plasmidsaurus.com/genome">service</a>. I sent the sample (~15mg cells suspended in Zymo DNA Shield) off on Tuesday morning and by Wednesday evening the results were ready.</p>
<p>Species ID (Mash)</p>
<ul>
<li>Best match: Pseudomonas putida NBRC 14164</li>
<li>Identity: 95.6% (1946/5000 shared hashes)</li>
</ul>
<p>Genome Quality (CheckM)</p>
<ul>
<li>Completeness: 99.88% â€” excellent!</li>
<li>Contamination: 2.15% â€” very low</li>
<li>Lineage marker: Pseudomonas</li>
</ul>
<p>Assembly Stats</p>
<ul>
<li>Genome size: 6.54 Mb (typical for Pseudomonas)</li>
<li>Total reads: 66,830 (398 Mb total)</li>
<li>Estimated coverage: ~58x</li>
<li>Longest read: 88.4 kb</li>
<li>Read N50: 9.7 kb</li>
</ul>
<p><img src="images/pseud_contig.png" class="img-fluid"></p>
<p>They give you <a href="https://drive.google.com/drive/folders/1p8dQGtGOjfowySz2ltuYEjhGZroxdjUh?usp=sharing">lots of data</a>, including annotating the genome for you with <a href="https://doi.org/10.1099/mgen.0.000685">bakta</a>. Great stuff! We can see right away that our guess was right - this is a Pseudomonas species (P. putida is the closest match) and are ready to dig in further to see what we can learn.</p>
</section>
<section id="pyoverdine-investigation" class="level2">
<h2 class="anchored" data-anchor-id="pyoverdine-investigation">Pyoverdine Investigation</h2>
<p>Pyoverdines vary but all come with three key parts: a dihydroxyquinoline core, a peptide chain, and a side chain. The peptide chain especially varies strain-to-strain - see <a href="https://en.wikipedia.org/wiki/Pyoverdine#Structure">here</a> for some examples.</p>
<p>Step one was looking for the pyoverdine Biosynthetic Gene Cluster (BGC). A <a href="https://chatgpt.com/share/697294ce-6a40-8010-bb23-a7047045b8cb">Deep Research</a> run gave some promising genes to start the hunt with. Some hits let us narrow in on a key region between 4.71 and 4.76M bp, with a number of promising-looking genes. A tool called antiSMASH likewise identified a region (4,696,705â€“4,796,912), labeled as â€˜NRP-metallophoreâ€™ - this is our BGC alright :)</p>
<p><img src="images/antiSMASH.png" class="img-fluid"></p>
<p>We also got a predicted structure in SMILES format, and with some wrangling get a final predicted structure (and a pretty picture):</p>
<p><img src="images/pyoverdine_precursor_pred.png" class="img-fluid"></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>AN AI SUMARY OF
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ ğŸ§ </p>
<section id="from-genome-to-siderophore-characterizing-a-novel-pyoverdine-from-pseudomonas-putida" class="level3">
<h3 class="anchored" data-anchor-id="from-genome-to-siderophore-characterizing-a-novel-pyoverdine-from-pseudomonas-putida">From Genome to Siderophore: Characterizing a Novel Pyoverdine from <em>Pseudomonas putida</em></h3>
</section>
<section id="the-starting-point" class="level3">
<h3 class="anchored" data-anchor-id="the-starting-point">The Starting Point</h3>
<p>We began with a high-quality assembled genome from a <em>Pseudomonas putida</em> strain isolated from a houseplant root system (sequenced by Plasmidsaurus). The assembly was excellent: <strong>99.88% complete</strong>, only 2.15% contamination, and a 6.54 Mb genome at ~58Ã— coverage. Species identity was confirmed via Mash (95.6% to <em>P. putida</em> NBRC 14164).</p>
</section>
<section id="finding-the-pyoverdine-biosynthetic-gene-cluster" class="level3">
<h3 class="anchored" data-anchor-id="finding-the-pyoverdine-biosynthetic-gene-cluster">Finding the Pyoverdine Biosynthetic Gene Cluster</h3>
<p>We searched the Bakta annotations for pyoverdine-related genes (<strong>pvd</strong> genes, NRPS, siderophore terms) and identified a cluster spanning <strong>~4.7â€“4.8 Mb</strong> containing:</p>
<ul>
<li><strong>pvdT</strong> â€” pyoverdine exporter</li>
<li><strong>Two large NRPS genes</strong> (~14.7 kb and ~6.4 kb) â€” the peptide assembly machinery</li>
<li><strong>PvdA-like monooxygenase</strong> â€” hydroxylates ornithine</li>
<li><strong>Formylglycine-generating enzyme</strong> (likely <strong>pvdF</strong>) â€” formyltransferase</li>
<li><strong>argD</strong> â€” makes diaminobutyrate (Dab), essential for the chromophore</li>
<li><strong>TonB-dependent receptor</strong> â€” likely the ferripyoverdine receptor (fpvA)</li>
</ul>
<p>Notably, <strong>pvdA</strong> was located separately at ~3.99 Mb â€” this split arrangement is common in pyoverdine BGCs.</p>
</section>
<section id="antismash-analysis" class="level3">
<h3 class="anchored" data-anchor-id="antismash-analysis">antiSMASH Analysis</h3>
<p>Running the genome through antiSMASH confirmed Region 10 as an <strong>NRP-metallophore</strong> cluster (siderophore NRPS). The top KnownClusterBlast hit was <strong>Pf-5 pyoverdine</strong> from <em>Pseudomonas protegens</em>, but with only ~49% NRPS identity â€” indicating a related but distinct peptide.</p>
</section>
<section id="predicting-the-peptide-sequence" class="level3">
<h3 class="anchored" data-anchor-id="predicting-the-peptide-sequence">Predicting the Peptide Sequence</h3>
<p>antiSMASH provided <strong>A-domain substrate predictions</strong> for each NRPS module:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Module</th>
<th>Substrate</th>
<th>Configuration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>Glu</td>
<td>L</td>
</tr>
<tr class="even">
<td>2</td>
<td>Tyr</td>
<td>D (E-domain present)</td>
</tr>
<tr class="odd">
<td>3</td>
<td>Dab</td>
<td>L</td>
</tr>
<tr class="even">
<td>4</td>
<td>Asp</td>
<td>L</td>
</tr>
<tr class="odd">
<td>5</td>
<td>Ala</td>
<td>L</td>
</tr>
<tr class="even">
<td>6</td>
<td>Asp</td>
<td>L</td>
</tr>
<tr class="odd">
<td>7</td>
<td>OH-Orn</td>
<td>D (E-domain present)</td>
</tr>
<tr class="even">
<td>8</td>
<td><strong>?</strong></td>
<td>â€”</td>
</tr>
<tr class="odd">
<td>9</td>
<td>OH-Orn</td>
<td>L</td>
</tr>
</tbody>
</table>
<p>The mystery at position 8 was annotated as <strong>â€œNHâ‚‚â€</strong> by the Minowa predictor â€” initially puzzling.</p>
</section>
<section id="resolving-the-unknown-residue" class="level3">
<h3 class="anchored" data-anchor-id="resolving-the-unknown-residue">Resolving the Unknown Residue</h3>
<p>We explored two hypotheses:</p>
<ol type="1">
<li><strong>Lysine</strong> â€” has a terminal -NHâ‚‚ on its side chain</li>
<li><strong>Glycine</strong> â€” essentially just an Î±-amino group with no real side chain</li>
</ol>
<p>After deeper analysis (including literature review of other <em>P. putida</em> pyoverdines), <strong>glycine emerged as the more likely candidate</strong>: - The â€œNHâ‚‚â€ prediction could mean â€œminimal/no side chainâ€ - Glycine is common in mid-chain positions of <em>P. putida</em> pyoverdines (KT2440, GB-1 both contain Gly) - Prediction tools are known to have difficulty with Gly vs Ala</p>
</section>
<section id="determining-the-side-chain-pvdn-vs-ptaa" class="level3">
<h3 class="anchored" data-anchor-id="determining-the-side-chain-pvdn-vs-ptaa">Determining the Side Chain: pvdN vs ptaA</h3>
<p>Pyoverdine side chains come in two main types: - <strong>Succinamide</strong> â€” produced by strains with <strong>pvdN</strong> - <strong>Î±-Ketoglutarate</strong> â€” produced by strains with <strong>ptaA</strong></p>
<p>We BLASTed known pvdN (<em>P. aeruginosa</em> PAO1, NP_251095.1) and ptaA sequences against the genome:</p>
<ul>
<li><strong>pvdN: 83% identity match</strong> â†’ gene LDMLEE_03614 at ~3.99 Mb (near pvdA)</li>
<li><strong>ptaA: No significant match</strong></li>
</ul>
<p><strong>Conclusion: Succinamide side chain</strong></p>
</section>
<section id="the-final-predicted-structure" class="level3">
<h3 class="anchored" data-anchor-id="the-final-predicted-structure">The Final Predicted Structure</h3>
<pre><code>CHROMOPHORE PRECURSORS          VARIABLE PEPTIDE
        â†“                              â†“
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”  â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
 L-Glu â†’ D-Tyr â†’ L-Dab â†’ L-Asp â†’ L-Ala â†’ L-Asp â†’ D-OHOrn â†’ Gly â†’ L-OHOrn
                  â†“                                    â†“            â†“
           cyclizes to form                      FeÂ³âº binding  FeÂ³âº binding
        dihydroxyquinoline                       (hydroxamate) (hydroxamate)
           chromophore</code></pre>
<p><strong>Key features:</strong> - <strong>9-residue peptide</strong> (longer than typical <em>P. putida</em> pyoverdines) - <strong>Two hydroxyornithine residues</strong> â€” provides two hydroxamate groups for FeÂ³âº chelation - <strong>Succinamide side chain</strong> on the chromophore - <strong>Novel sequence</strong> â€” no match in SIDERITE database (best similarity only 0.35)</p>
</section>
<section id="smiles-with-stereochemistry" class="level3">
<h3 class="anchored" data-anchor-id="smiles-with-stereochemistry">SMILES (with stereochemistry)</h3>
<pre><code>NN[C@@H](CCC(=O)O)C(=O)N[C@H](Cc1ccc(O)cc1)C(=O)N[C@@H](CCN)C(=O)N[C@@H](CC(=O)O)C(=O)N[C@@H](C)C(=O)N[C@@H](CC(=O)O)C(=O)N[C@H](CCCNO)C(=O)NCC(=O)N[C@@H](CCCNO)C(=O)O</code></pre>
</section>
<section id="what-still-needs-confirmation" class="level3">
<h3 class="anchored" data-anchor-id="what-still-needs-confirmation">What Still Needs Confirmation</h3>
<p>This is a <strong>bioinformatic prediction</strong>. To definitively confirm the structure:</p>
<ol type="1">
<li><strong>LC-MS/MS</strong> â€” will reveal the exact mass and fragment pattern, confirming each residue (especially Gly at position 8)</li>
<li><strong>The hydroxamates</strong> â€” could be N-formylated (if pvdF is active), which would add +28 Da per residue</li>
<li><strong>Cyclization</strong> â€” the C-terminal OHOrn likely cyclizes with the chromophore, which MS/MS should reveal</li>
</ol>
</section>
<section id="why-this-matters" class="level3">
<h3 class="anchored" data-anchor-id="why-this-matters">Why This Matters</h3>
<p>This appears to be a <strong>genuinely novel pyoverdine variant</strong> â€” it doesnâ€™t match any characterized <em>P. putida</em> pyoverdines in the literature or databases. Discovering the diversity of siderophores in environmental isolates helps us understand:</p>
<ul>
<li>Microbial competition for iron in the rhizosphere</li>
<li>The evolution of NRPS systems</li>
<li>Potential for novel iron chelators with biotechnology applications</li>
</ul>
</section>
<section id="end-ai-slop" class="level2">
<h2 class="anchored" data-anchor-id="end-ai-slop">END AI SLOP</h2>
</section>
</div>
</div>
</div>
<p>Hereâ€™s <a href="https://gist.github.com/johnowhitaker/8659d6f38c799e9d4d8e417e1641e6e0">a notebook</a> running over the final steps, with some light annotation.</p>
</section>
<section id="future-plans" class="level2">
<h2 class="anchored" data-anchor-id="future-plans">Future Plans</h2>
<p>Itâ€™s one thing to poke at a genome and predict a structure, itâ€™s another entirely to verify it. Iâ€™m hoping to work with a local university to run GC-MS on this to answer a few remaining questions and nail down the final structure exactly. Will require dusting off and improving my organic chemistry and learning a lot more! Stay tuned for that in some future blog post hopefully :)</p>
<p>Also, itâ€™s probably obvious but worth stating explicityl: I AM OUT OF MY DEPTH HERE AND EVERYTHING IN THIS POST SHOULD BE TAKEN WITH A GRAIN OF SALT, ESPECIALLY WHILE IT IS MARKED â€˜DRAFTâ€™ :)</p>
<p>For the curious, raw data <a href="https://drive.google.com/drive/folders/1p8dQGtGOjfowySz2ltuYEjhGZroxdjUh?usp=sharing">is on Google Drive</a>, Iâ€™m open to questions <span class="citation" data-cites="johnowhitaker">@johnowhitaker</span>.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/johnowhitaker\.dev");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>